### ISSUE TITLE
WYSIWYG Math editor

### ISSUE THREAD

**Issue by Carreau** [jupyter/notebook#1604]
Which emits latex:

https://github.com/mathquill/mathquill

I would **strongly** recommend to try the demo:

http://mathquill.com/

cc @jasongrout , @SylvainCorlay , @ellisonbg , @fperez 

If we could have an inputhook in md cells to enter math formula this way, that would be an extremely nice feature.

**Comment by fperez** [comment]
Very cool tool, though for me the best of this kind remains [LyX](http://www.lyx.org). It lets you combine typing LaTeX that gets auto-rendered (like MathQuill) with having command palettes for more esoteric stuff you may not know/remember how to type in LaTeX.  

That makes it incredibly powerful: if you know the commands, you just type along and it simply "compiles on the fly" to WYSIWYM ("What you see is what you mean", i.e. close enough visually to the final LaTeX render to be useful, but not pixel-perfect so it can be done on the fly without a full LaTeX call). But if you don't know something, you can use the palettes to visually create the piece you are missing.  When I was doing lots more complex mathematical writing, it saved me endless amounts of poring over LaTeX docs, and I even used it often to answer colleagues' questions about how to do something in LaTeX: I would just open LyX, type/build up what they needed, and then view the raw LaTeX (which it will show you if you want) for the answer to copy/paste into a raw `.tex` file.

I've yet to find a system for math editing that I like better than LyX.

**Comment by Carreau** [comment]
You can see the HackerNew Thread :https://news.ycombinator.com/item?id=12065699 , arguably the command palette part is the easy part, as lomg as you have WYSIWYG movement end rendering, adding buttons that insert the right latex seem like the easy part. 

[This](https://www.codecogs.com/latex/eqneditor.php) has a palette but the edition/redering are in 2 different area.

Having used LyX, I think http://mathquill.com/ is the closest to something that could eventually replace it that I have seen.

**Comment by ellisonbg** [comment]
This is truly amazing! I wrote my Ph.D. using LyX and it would be truly
incredible to have this in the notebook.

Cheers,

Brian

On Sun, Jul 10, 2016 at 5:20 PM, Matthias Bussonnier <
notifications@github.com> wrote:

> You can see the HackerNew Thread :
> https://news.ycombinator.com/item?id=12065699 , arguably the command
> palette part is the easy part, as lomg as you have WYSIWYG movement end
> rendering, adding buttons that insert the right latex seem like the easy
> part.
>
> This <https://www.codecogs.com/latex/eqneditor.php> has a palette but the
> edition/redering are in 2 different area.
>
> Having used LyX, I think http://mathquill.com/ is the closest to
> something that could eventually replace it that I have seen.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jupyter/notebook/issues/1604#issuecomment-231614347>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AABr0C6OF6GZCMkP3ot-KagSxYrsxnALks5qUXBBgaJpZM4JI4Ln>
> .
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by topisani** [comment]
yep, this would be absolutely amazing, i might take a look at it

**Comment by ragloo** [comment]
+1
Is there any move/decision towards this direction ?

**Comment by terryphi** [comment]
+1 
would be very useful.  Mathquill implementation would be great.

**Comment by ranr01** [comment]
+1 
I just thought about it myself... Nice to see someone already did.

**Comment by JustinGirard** [comment]
Did this ever get integrated?

**Comment by jasonsun0310** [comment]
+1. I think this will be very useful

**Comment by doppioslash** [comment]
+1 
this would be good, any news?

**Comment by skirpichev** [comment]
> If we could have an inputhook in md cells to enter math formula this way

It might be useful to have inputhook for code cells as well.  This may enable interactive editing of math in CAS like SymPy or Diofant, as the Mathematica currently does.

**Comment by jasonsun0310** [comment]
Just a quick question: if jupyter console is ported into LyX, how many of you will be happy?

**Comment by ragloo** [comment]
I will :+1:

**Comment by jasonsun0310** [comment]
how about something like this?

![Screen Shot 2019-07-02 at 12 14 15 PM](https://user-images.githubusercontent.com/22034141/60528729-1a946780-9cc3-11e9-8a2c-ae70316a88e8.jpg)

**Comment by ragloo** [comment]
Love it !!!
Awesome work :+1:

**Comment by meeseeksmachine** [comment]
This issue has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/does-jupyter-notebook-have-a-wysiwyg-math-editor/11469/2

**Comment by deronj** [comment]
I tried using Mathquill and it has a very nice feel. It imilar to LyX in the way you can use the arrow keys to move to various parts of the equation. However, I think the highlighting of the current focus area could be better. If I recall, LyX uses an outline box to indicate the focus area. In contrast, mathquill uses a blinking vertical cursor, which in my opinion is less visible than the outline box. Of course, a more complicated implementation could allow the user to customize the highlighting method.

Mathquill could provide some leverage for implementing a WYSIWYG math editor in Jupyter because it has a relatively permissive license (MPL). LyX, on the other hand, is GPL2. This is incompatible with the Jupyter BSD-3-clause license.

**Comment by fperez** [comment]
Thanks for this pointer @deronj! I love LyX (have used it since the 90's!) but not only is the GPL an issue as you correctly point out, their implementation is C++/Qt, targeted at the desktop. I can't imagine that code being much use to us, as we need a web-based solution. But Mathquill _is_ a web-native tool, so along with a compatible license, this sounds like a great option to look at.

**Comment by fperez** [comment]
Duh - deleting my comment above - I'd "reset cache" on this thread about lyx and forgot it _started_ with mathquill.. 🤦

**Comment by spchamp** [comment]
Proceeding from an idea of adding support for editing math expressions in the LaTeX-in-markdown support in jupyter notebooks, perhaps for using [ckeditor with react support](https://ckeditor.com/docs/ckeditor5/latest/getting-started/installation/react/react.html#quick-start) and [ckeditor5-math](https://github.com/isaul32/ckeditor5-math) in jupyter, one might be able to extend any of the existing [extension points](https://jupyterlab.readthedocs.io/en/4.2.x/extension/extension_points.html) for JupyterLab on the JS side. I'm not certain of how this could be ported over to Jupyter Notebooks, if absent of JupyterLab.

There are a couple of [walk-throughs](https://jupyterlab.readthedocs.io/en/4.2.x/extension/notebook.html#how-to-extend-the-notebook-plugin) available for adding notebook extensions. The [widget walk-through](https://jupyterlab.readthedocs.io/en/4.2.x/extension/notebook.html#how-to-extend-the-notebook-plugin) shows how to use the [_copier_ extension-template](https://github.com/jupyterlab/extension-template) when adding an extension using typescript. In the example, perhaps the `WidgetExtension` class might be something to consider, if one could add ckeditor support via an `IWidgetExtension`. 

For integrating with `$inline$` and `$$block$$` math expressions in Jupyter markdown handling, perhaps the extension could use the cell model and cell events in Jupyter, somehow? Taking a look at the JupyterLab sources, there are the [`MarkdownCell`](https://github.com/jupyterlab/jupyterlab/blob/8e67004ae446fea3ffaeec561e47d855f8847f01/packages/cells/src/widget.ts#L2105) and [`MarkdownCellModel`](https://github.com/jupyterlab/jupyterlab/blob/8e67004ae446fea3ffaeec561e47d855f8847f01/packages/cells/src/model.ts#L576) classes on the JS side. There's also the CSS class model for how markdown cells are rendered in a [Jupyter notebook widget](https://jupyterlab.readthedocs.io/en/4.2.x/extension/notebook.html#notebook-widget).  

When proposing an approach with reactpy, I wasn't aware of this existing framework in jupyter. It seems that there's a lot that could be extended here, on the JS side.

For extensions in python, of course there's also [ipywidgets](https://ipywidgets.readthedocs.io/)

**Comment by fperez** [comment]
cc @rowanc1 @agoose77 as I know the MyST team has also thought about this...
